src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,406 @@
+
+  /** Show friendly type name with current scope in mind
+   *
+   *  E.g.    C.this.B     -->  B     if current owner is C
+   *          C.this.x.T   -->  x.T   if current owner is C
+   *          X[T]         -->  X
+   *          C            -->  C     if current owner is C !!!
+   *
+   */
+  def showType(tp: Type): String = {
+    val enclosingCls = ctx.owner.enclosingClass.asClass.classInfo.symbolicTypeRef
+
+    def isOmittable(sym: Symbol) =
+      sym.isEffectiveRoot || sym.isAnonymousClass || sym.name.isReplWrapperName ||
+        ctx.definitions.UnqualifiedOwnerTypes.exists(_.symbol == sym) ||
+        sym.showFullName.startsWith("scala.") ||
+        sym == enclosingCls.typeSymbol
+
+    def refinePrefix(tp: Type): String = tp match {
+      case NoPrefix => ""
+      case tp: NamedType if isOmittable(tp.symbol) => ""
+      case tp: ThisType => refinePrefix(tp.tref)
+      case tp: RefinedType => refinePrefix(tp.parent)
+      case tp: NamedType => tp.name.show.stripSuffix("$")
+    }
+
+    def refine(tp: Type): String = tp match {
+      case tp: RefinedType => refine(tp.parent)
+      case tp: ThisType => refine(tp.tref)
+      case tp: NamedType =>
+        val pre = refinePrefix(tp.prefix)
+        if (pre.isEmpty) tp.name.show.stripSuffix("$")
+        else pre + "." + tp.name.show.stripSuffix("$")
+      case _ => tp.show.stripSuffix("$")
+    }
+
+    val text = tp match {
+      case tp: OrType => showType(tp.tp1) + " | " + showType(tp.tp2)
+      case _ => refine(tp)
+    }
+
+    if (text.isEmpty) enclosingCls.show.stripSuffix("$")
+    else text
+  }
+
+  /** Display spaces */
+  def show(s: Space, mergeList: Boolean = false): String = s match {
+    case Empty => ""
+    case Typ(tp, decomposed) =>
+      val sym = tp.widen.classSymbol
+
+      if (sym.is(ModuleClass))
+        showType(tp)
+      else if (ctx.definitions.isTupleType(tp))
+        "(" + signature(tp).map(_ => "_").mkString(", ") + ")"
+      else if (sym.showFullName == "scala.collection.immutable.::")
+        if (mergeList) "_" else "List(_)"
+      else if (tp.classSymbol.is(CaseClass))
+        // use constructor syntax for case class
+        showType(tp) + "(" + signature(tp).map(_ => "_").mkString(",") + ")"
+      else if (signature(tp).nonEmpty)
+        tp.classSymbol.name + "(" + signature(tp).map(_ => "_").mkString(",") + ")"
+      else if (decomposed) "_: " + showType(tp)
+      else "_"
+    case Kon(tp, params) =>
+      if (ctx.definitions.isTupleType(tp))
+        "(" + params.map(p => show(p)).mkString(", ") + ")"
+      else if (tp.widen.classSymbol.showFullName == "scala.collection.immutable.::")
+        if (mergeList) params.map(p => show(p, mergeList)).mkString(", ")
+        else "List(" + params.map(p => show(p, true)).mkString(", ") + ")"
+      else
+        showType(tp) + "(" + params.map(p => show(p)).mkString(", ") + ")"
+    case Or(spaces) => "{ " + spaces.map(p => show(p)).mkString(", ") + " }"
```

Comment:
The default implementation of `Set` does not have a stable ordering, if you don't want your messages to be non-deterministic you should sort the spaces before printing them.